### PR TITLE
feat(ast): optimize node memory layouts

### PR DESCRIPTION
**Pull request by sno2**
| Node     | Bytes Before | Bytes After |
|----------|--------------|-------------|
| E.String | 48           | 40          | 
| E.Local  | 32           | 24          | 

**Hold on review please - going to add more optimizations.

### CODE REVIEW

File: src/js_ast.zig [hunk 1]
```diff
@@ -1615,15 +1615,18 @@ pub const E = struct {
...
 
         // A very simple rope implementation
         // We only use this for string folding, so this is kind of overkill
         // We don't need to deal with substrings
         next: ?*String = null,
         end: ?*String = null,
         rope_len: u32 = 0,
-        is_utf16: bool = false,
+
+        flags: packed struct {
```
Comment (Jarred-Sumner) [inline src/js_ast.zig:1626]: `packed` is super buggy in zig's stage1 compiler. Let's use `EnumStructField` instead like the other `Flags` types in `js_ast.zig`

Comment (Jarred-Sumner) [inline src/js_ast.zig:1626]: `packed` is super buggy in zig's stage1 compiler. Let's use `EnumStructField` instead like the other `Flags` types in `js_ast.zig`

**Comment by Jarred-Sumner** [comment]
Reducing the size of these is nice, but note that it likely won't have impact unless the size of the largest struct is reduced. Let's avoid increasing code complexity unless you see a % impact here to transpilation time (easiest to measure by copy-pasting `@babel/standalone` a few times in one file and running `bun build` on it.

The reason why is `BaseStore`. Memory allocated for the AST nodes is the largest of the `E.Data` union types. We need to reduce the size of the biggest of those to reduce the size of memory allocated overall.

**Comment by sno2** [comment]
> The reason why is BaseStore. Memory allocated for the AST nodes is the largest of the E.Data union types. We need to reduce the size of the biggest of those to reduce the size of memory allocated overall.

Hmm I noticed `E.Data` and I figured they were all smaller than or equal to a pointer so you allocated it separately.  Is there a specific reason why you did not allocate the data as the size needed for each tag kind?

Edit: Oh, it seems for faster iterations over the data I guess?

**Comment by Jarred-Sumner** [comment]
https://github.com/Jarred-Sumner/bun/blob/0b1d4ff81bbb19383cba7932e88359a6ccfb6899/src/js_ast.zig#L31-L182

it fits in preallocated arrays like this which was a meaningfully large performance improvement

**Comment by sno2** [comment]
This does not seem worth it after understanding the points that Jarred stated.

**Pull request closed without merging**